texar/modules/connectors/connectors.py

```diff
@@ -0,0 +1,866 @@
+
+    @staticmethod
+    def default_hparams() -> dict:
+        r"""Returns a dictionary of hyperparameters with default values.
+
+        .. code-block:: python
+
+            {
+                "activation_fn": "texar.core.layers.identity",
+                "name": "reparameterized_stochastic_connector"
+            }
+
+        Here:
+
+        "activation_fn" : str
+            The activation function applied to the outputs of the MLP
+            transformation layer. Can be a function,
+            or its name or module path.
+        "name" : str
+            Name of the connector.
+        """
+        return {
+            "activation_fn": "texar.core.layers.identity",
+            "name": "reparameterized_stochastic_connector"
+        }
+
+    def forward(self,    # type: ignore
+                num_samples: Optional[Union[int, torch.Tensor]] = None,
+                transform: bool = True) -> Tuple[Any, Any]:
+        r"""Samples from a distribution and optionally performs transformation
+        with an MLP layer.
+        The distribution must be reparameterizable, i.e.,
+        :python:`Distribution.has_rsample == True`.
+
+        Args:
+            num_samples (optional): An ``int`` or ``int`` ``Tensor``.
+                Number of samples to generate. If not given,
+                generate a single sample. Note that if batch size has
+                already been included in :attr:`distribution`'s dimensionality,
+                :attr:`num_samples` should be left as ``None``.
+            transform (bool): Whether to perform MLP transformation of the
+                distribution samples. If ``False``, the structure/shape of a
+                sample must match :attr:`output_size`.
+
+
+        :returns:
+            If attr:`transform` is ``True``, returns a ``tuple``
+            (attr:`output`, attr:`sample`), where
+            - output: A ``Tensor`` or a (nested) ``tuple`` of Tensors with
+              the same structure and size of :attr:`output_size`.
+              The batch dimension equals :attr:`num_samples` if specified,
+              or is determined by the distribution dimensionality.
+            - sample: The sample from the distribution, prior to transformation.
+
+            Otherwise, returns a ``Tensor`` attr:`sample`, where
+            - sample: The sample from the distribution, prior to transformation.
+
+        Raises:
+            ValueError: If distribution cannot be reparametrized.
+            ValueError: The output does not match :attr:`output_size`.
+        """
+
+        distribution = utils.check_or_get_instance(
+            self._dstr_type, self._dstr_kwargs,
+            ["torch.distributions", "torch.distributions.multivariate_normal",
+             "texar.custom"])
+
+        assert distribution.has_rsample    # type: ignore
+
+        if num_samples:
+            sample = distribution.rsample([num_samples])    # type: ignore
```

Comment:
Why do we need a `# type: ignore` comment? Is it possible to declare a type for `sample` that does not clash with the two assignments here?

Reply:
Yeah... There will be error "error: `T" has no attribute "rsample"` if not `# type: ignore`.

Reply:
I see. I think the problem here is that we should not use a generic type `T` but rather just the `torch.Distribution` type. Could you try to see if this gets rid of the error?